tempesta_fw/http_parser.c

```diff
@@ -1031,37 +1032,40 @@ __parse_transfer_encoding(TfwHttpMsg *hm, unsigned char *data, size_t len)
 	 * According to RFC 7230 section 3.3.1:
 	 *
 	 * TODO: In a response:
-	 * A server MUST NOT send a Transfer-Encoding header field in any
-	 * response with a status code of 1xx (Informational) or 204 (No
-	 * Content).  A server MUST NOT send a Transfer-Encoding header
-	 * field in any 2xx (Successful) response to a CONNECT request.
-	 *
-	 * TODO: A sender MUST NOT apply chunked more than once to a message
-	 * body (i.e., chunking an already chunked message is not allowed).
-	 *
-	 * TODO: in a request:
-	 * If any transfer coding other than chunked is applied to a request
-	 * payload body, the sender MUST apply chunked as the final transfer
-	 * coding to ensure that the message is properly framed.
-	 *
-	 * In a response, it's permissible to not have chunked transfer
-	 * coding. In that case the message is terminated by closing the
-	 * connection.
+	 * A server MUST NOT send a Transfer-Encoding header field
+	 * in any 2xx (Successful) response to a CONNECT request.
 	 */
 	__FSM_STATE(I_TransEncod) {
+		if (TFW_CONN_TYPE(hm->conn) & Conn_Srv) {
+			unsigned int status = ((TfwHttpResp *)hm)->status;
+			if ((status - 100U < 100U) || (status == 204))
+				return CSTR_NEQ;
+		}
+		__FSM_I_JMP(I_TransEncodChunked);
+	}
+
+	__FSM_STATE(I_TransEncodChunked) {
+		/*
+		 * A sender MUST NOT apply chunked more than once
+		 * to a message body (i.e., chunking an already
+		 * chunked message is not allowed). RFC 7230 3.3.1.
+		 */
 		TRY_STR_LAMBDA("chunked", {
-			msg->flags |= TFW_HTTP_CHUNKED;
+			if (unlikely(msg->flags & TFW_HTTP_CHUNKED))
+				return CSTR_NEQ;
+			msg->flags |= TFW_HTTP_CHUNKED
+				      | __TFW_HTTP_CHUNKED_LAST;
 		}, I_EoT);
 		TRY_STR_INIT();
-		__FSM_I_MOVE_n(I_TransEncodExt, 0);
+		__FSM_I_MOVE_n(I_TransEncodOther, 0);
 	}
 
-	__FSM_STATE(I_TransEncodExt) {
+	__FSM_STATE(I_TransEncodOther) {
```

Comment:
Every time we reach this state when `TFW_HTTP_CHUNKED` is already set we can invalidate message: chunked encoding is not the last encoding. And we don't need to use `__TFW_HTTP_CHUNKED_LAST`flag